iree/compiler/Dialect/Stream/Transforms/RefineUsage.cpp

```diff
@@ -0,0 +1,334 @@
+    return didChange ? success() : failure();
+  }
+};
+
+// Applies usage analysis results to an MLIR function.
+// All resource arguments and results, block arguments, and nested operations
+// will have their lifetime specified.
+struct ApplyFuncOp : public UsageRefinementPattern<mlir::FuncOp> {
+  using UsageRefinementPattern<mlir::FuncOp>::UsageRefinementPattern;
+  LogicalResult matchAndRewrite(mlir::FuncOp op,
+                                PatternRewriter &rewriter) const override {
+    bool didChange = false;
+
+    // Arguments:
+    SmallVector<Type> newInputs;
+    for (auto inputType : llvm::enumerate(op.getType().getInputs())) {
+      auto oldType = inputType.value().dyn_cast<IREE::Stream::ResourceType>();
+      if (!oldType) {
+        newInputs.push_back(inputType.value());
+        continue;
+      }
+      auto blockArg = op.getArgument(inputType.index());
+      auto newUsage = analysis.lookupResourceUsage(blockArg);
+      auto newLifetime = convertUsageToLifetime(newUsage);
+      auto newType = rewriter.getType<IREE::Stream::ResourceType>(newLifetime);
+      newInputs.push_back(newType);
+    }
+
+    // Results:
+    SmallVector<Type> newOutputs;
+    auto anyReturnOp = *op.getOps<mlir::ReturnOp>().begin();
+    for (auto outputType : llvm::enumerate(op.getType().getResults())) {
+      auto oldType = outputType.value().dyn_cast<IREE::Stream::ResourceType>();
+      if (!oldType) {
+        newOutputs.push_back(outputType.value());
+        continue;
+      }
+      auto returnValue = anyReturnOp.getOperand(outputType.index());
+      auto newUsage = analysis.lookupResourceUsage(returnValue);
+      auto newLifetime = convertUsageToLifetime(newUsage);
+      auto newType = rewriter.getType<IREE::Stream::ResourceType>(newLifetime);
+      newOutputs.push_back(newType);
+    }
+    auto newFuncType = rewriter.getFunctionType(newInputs, newOutputs);
+    if (op.getType() != newFuncType) {
+      op.setType(newFuncType);
+      didChange = true;
+    }
+
+    // Blocks and nested operations:
+    didChange = this->applyRegionTransitions(op, rewriter) || didChange;
+    return didChange ? success() : failure();
+  }
+};
+
+// Applies usage analysis results to a generic MLIR op.
+// All resource operands and results including those in nested regions will have
+// their lifetime specified.
+template <typename Op>
+struct ApplyGenericOp : public UsageRefinementPattern<Op> {
+  using UsageRefinementPattern<Op>::UsageRefinementPattern;
+  LogicalResult matchAndRewrite(Op op,
+                                PatternRewriter &rewriter) const override {
+    bool didChange = this->applyRegionTransitions(op, rewriter);
+    rewriter.startRootUpdate(op);
+    for (unsigned i = 0; i < op->getNumResults(); ++i) {
+      auto result = op->getResult(i);
+      if (!result.getType().template isa<IREE::Stream::ResourceType>())
+        continue;
+      didChange = this->applyResultTransition(op, result, rewriter);
```

Comment:
I think you need this to be:

```
if (this->applyResultTransition(..))
  didChange = true;
```

(or use `|=` which always sketches me out a bit because I get confused about logical vs bitwise on bool with short circuit eval)